vividus-plugin-web-app/src/main/java/org/vividus/proxy/ThreadedProxy.java

```diff
@@ -16,73 +16,157 @@
 
 package org.vividus.proxy;
 
-import javax.inject.Inject;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.function.Supplier;
 
 import com.browserup.bup.BrowserUpProxy;
 import com.browserup.bup.filters.RequestFilter;
 
+import org.apache.commons.lang3.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.model.IntegerRange;
+
 public class ThreadedProxy implements IProxy
 {
-    @Inject private IProxyServerFactory proxyServerFactory;
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThreadedProxy.class);
+
+    private final InetAddress proxyHost;
+    private final Queue<Integer> proxyPorts;
+    private final boolean useEphemeralPort;
+    private final ThreadLocal<IProxy> proxy;
+
+    public ThreadedProxy(String proxyHost, IntegerRange proxyPorts, IProxyFactory proxyFactory)
+            throws UnknownHostException
+    {
+        List<Integer> proxyPortsRange = proxyPorts.getRange();
+        if (proxyPortsRange.contains(0))
+        {
+            Validate.isTrue(proxyPortsRange.size() == 1,
+                    "Port 0 (ephemeral port selection) can not be used with custom ports");
+            useEphemeralPort = true;
+        }
+        else
+        {
+            useEphemeralPort = false;
+            proxyPortsRange.forEach(ThreadedProxy::validatePort);
+        }
+        this.proxyPorts = new LinkedList<>(proxyPortsRange);
+        this.proxy = ThreadLocal.withInitial(proxyFactory::createProxy);
+        this.proxyHost = InetAddress.getByName(proxyHost);
+    }
 
-    private final ThreadLocal<IProxy> proxy = ThreadLocal.withInitial(() ->
+    @SuppressWarnings("MagicNumber")
+    private static void validatePort(int port)
     {
-        Proxy proxy = new Proxy();
-        proxy.setProxyServerFactory(proxyServerFactory);
-        return proxy;
-    });
+        Validate.isTrue(port > 0 && port <= 65535, "Expected ports range is 1-65535 but got: %d", port);
+    }
 
     @Override
     public void start()
     {
-        proxy.get().start();
+        perform(() -> proxy().start(), () ->
+        {
+            Validate.isTrue(!proxyPorts.isEmpty(), "There are no available ports in the ports pool");
+            int port = proxyPorts.poll();
+            logMessage("Allocate {} port from the proxy ports pool", () -> port);
```

Comment:
`LOGGER.atInfo().addArgument(port).log("Allocate {} port from the proxy ports pool");`